src/v/utils/tracing_mutex.h

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+#include "seastarx.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/semaphore.hh>
+
+#include <utility>
+
+/*
+ * This is a extension of mutex.h. The primary motivation for
+ * this class is to gather statistics on how ofter the mutex is
+ * taken (lock_counter) and the id of the top lock holder.
+ *
+ * Usage
+ * =====
+ *
+ *    ```
+ *    enum holder_id { tracing_mutex_example = 0 }
+ *
+ *    mutex m;
+ *    return m.with(holder_id::tracing_mutex_example, [] { ... });
+ *    ```
+ *
+ */
+
+class tracing_mutex;
+
+class locked_token {
+    tracing_mutex* _mutex;
+    ss::semaphore_units<> _unit;
+    bool _is_moved{false};
+    int _lock_holder_id;
+
+public:
+    locked_token(
+      tracing_mutex* mutex,
+      int lock_holder_id,
+      ss::semaphore_units<>&& unit) noexcept
+      : _mutex(mutex)
+      , _lock_holder_id(lock_holder_id)
+      , _unit(std::move(unit)) {}
+    locked_token(locked_token&& token) noexcept
+      : _mutex(token._mutex)
+      , _unit(std::move(token._unit)) {
+        token._is_moved = true;
+    }
+    locked_token(const locked_token&) = delete;
```

Comment:
With optional you just need to make c-tors 'default' or 'delete' and I'd suggest to add assignment and move-assignment operators here too.